app/app.py

```diff
@@ -85,49 +88,69 @@ def extract_signature_info(image_path):
 # FUZZY MATCHING FUNCTION
 ##
 
-def score_function_fuzz(ocr_name, full_name_list):
-
-    """
-    Outputs the voter record indices of the names that are 
-    closest to `ocr_name`.
-    """
-
-    # empty dictionary of scores 
-    full_name_score_dict = dict()
-
-    for idx in range(len(full_name_list)):
-
-        # getting full name for row; ensuring string
-        name_row = str(full_name_list[idx])
+def score_fuzzy_match_slim(ocr_name, full_name_list, scorer_=fuzz.token_ratio, limit_=1):
+    list_of_match_tuples = process.extract(query=ocr_name, choices=full_name_list, scorer=scorer_, processor=utils.default_process, limit=limit_)
+    return list_of_match_tuples
 
-        # converting string to lower case to simplify matching  
-        name_row = name_row.lower()
-        ocr_name = ocr_name.lower()
-    
-        # compiling scores; writing as between 0 and 1
-        full_name_score_dict[idx] = fuzz.ratio(ocr_name, name_row)/100
 
-    # sorting dictionary
-    sorted_dictionary = dict(sorted(full_name_score_dict.items(), reverse=True, key=lambda item: item[1]))
-
-    # top five key value pairs (indices and scores)
-    indices_scores_list = list(sorted_dictionary.items())[:5]
+##
+# TIERED SEARCH
+##
 
-    return indices_scores_list
+def tiered_search(name, address):
```

Comment:
The `if-else` conditionals in this function are confusingly written. I think the logic is OK but it is hard to parse. 

I tried to rewrite it to be more readable. Let me know if you think the below function does something different

```
def tiered_search(name, address):

    name_address_combo = f"{name} {address}"

    # Searches for a match within the Ward returned by OCR
    name_address_matches = score_fuzzy_match_slim(name_address_combo, voter_records_2023_df[voter_records_2023_df['WARD'] == f"{dict_['Ward']}.0"]["Full Name and Full Address"])
    name_address_name, name_address_score, name_address_id = name_address_matches[0]

    # if score is more than 85, return the tuple
    if name_address_score >= 85.0:
        return (name_address_name, name_address_score, name_address_id)
    
    # if score is below 85, do additional processing
    else:

        # computing matches based on name and address; only considers all other wards
        name_address__matches = score_fuzzy_match_slim(name_address_combo, voter_records_2023_df[voter_records_2023_df['WARD'] != f"{dict_['Ward']}.0"]["Full Name and Full Address"])
        name_address__name, name_address__score, name_address__id = name_address__matches[0]

        # computing matches based on name alone; considers full voter records
        full_name__matches = score_fuzzy_match_slim(name, voter_records_2023_df["Full Name"], scorer_=fuzz.ratio)
        full_name__name, full_name__score, full_name__id = full_name__matches[0]

        # if score based on name and address is greater than score based on full name, then output name and address info
        if name_address__score >= full_name__score:
            return (name_address__name, name_address__score, name_address__id)

        # if name address score is less than full name, then output full name score
        # but first search for close address in voter records
        else:
            address = voter_records_2023_df.loc[full_name__id, 'Full Address']
            full_name_address = f"{full_name__name} {address}"
            return (full_name_address, full_name__score, full_name__id)
```

Reply:
This version functions the same assuming there is a match, but the old version will keep the record with the highest score between the name_address ward search and the name_address registry search. This only matters if none of the searches produces a valid match, and the Ward search returns a higher match % than the Full name search, which may be statistically insignificant. In any case it definitely won't affect valid match % any, just the highest result return when there are no valid matches. I'll update it to use this version,

Reply:
Ah I see. I missed that case: It's possible that the ward-based score is less than 85, but that the scores for the full registry are less than the ward-based score in which case we would want to return the ward-based results. I think we can encompass this case by slightly rewriting the function. Here is a new version. Let me know if it misses any other cases from the original

```
def tiered_search(name, address):

    name_address_combo = f"{name} {address}"

    # Searches for a match within the Ward returned by OCR
    name_address_matches1 = score_fuzzy_match_slim(name_address_combo, voter_records_2023_df[voter_records_2023_df['WARD'] == f"{dict_['Ward']}.0"]["Full Name and Full Address"])
    name_address__name1, name_address__score1, name_address__id1 = name_address_matches1[0]

    # if score is more than 85, return the tuple
    if name_address__score1 >= 85.0:
        return (name_address__name1, name_address__score1, name_address__id1)
    
    # if score is below 85, do additional processing
    else:

        # computing matches based on name and address; only considers all other wards
        name_address_matches2 = score_fuzzy_match_slim(name_address_combo, voter_records_2023_df[voter_records_2023_df['WARD'] != f"{dict_['Ward']}.0"]["Full Name and Full Address"])
        name_address__name2, name_address__score2, name_address__id2 = name_address_matches2[0]

        # computing matches based on name alone; considers full voter records
        full_name_matches = score_fuzzy_match_slim(name, voter_records_2023_df["Full Name"], scorer_=fuzz.ratio)
        full_name__name, full_name__score, full_name__id = full_name_matches[0]

        # if original ward-based score is the highest, then return ward-based results
        if (name_address__score1 >= name_address__score2) and (name_address__score1 >= full_name__score):
            return (name_address__name1, name_address__score1, name_address__id1)

        else:
            # if score based on name and address is greater than score based on full name, then output name and address info
            if name_address__score2 >= full_name__score:
                return (name_address__name2, name_address__score2, name_address__id2)

            # if name address score is less than full name, then output full name score
            # but first search for address in voter records
            else:
                address = voter_records_2023_df.loc[full_name__id, 'Full Address']
                full_name_address = f"{full_name__name} {address}"
                return (full_name_address, full_name__score, full_name__id)
```

Reply:
I rewrote this slightly, only so we don't have to traverse the full dataframe twice there's a match found in the Full Name + Address registry search. Let me know what you think about the readability.

    def tiered_search(name, address):
    name_address = f"{name} {address}"

    # Searches for a match within the Ward returned by OCR

    name_address_matches = score_fuzzy_match_slim(name_address, voter_records_2023_df[voter_records_2023_df['WARD'] == f"{dict_['Ward']}.0"]["Full Name and Full Address"])
    name_address__name, name_address__score, name_address__id = name_address_matches[0]

    # If no Valid matches are found in the ward, searches for a match against the entire registry except the ward already checked

    if name_address__score < 85.0:
        name_address_matches2 = score_fuzzy_match_slim(name_address, voter_records_2023_df[voter_records_2023_df['WARD'] != f"{dict_['Ward']}.0"]["Full Name and Full Address"])
        name_address__score2 = name_address_matches2[0][1]

        # If the Registry search is higher than the Ward search, reassigns values for score check and Full Name score comparison

        if name_address__score < name_address__score2:
            name_address__name, name_address__score_, name_address__id = name_address_matches2[0]

    # If there is a valid match on either search, returns the record
    if name_address__score >= 85.0:
        return (name_address__name, name_address__score, name_address__id)

    # IF no Valid matches have been found, searches for a match using only the Full Name

    else:
        full_name_matches = score_fuzzy_match_slim(name, voter_records_2023_df["Full Name"], scorer_=fuzz.ratio)
        full_name__name, full_name__score, full_name__id = full_name_matches[0]

    # Compare scores of full name + address match to score of Full Name match and take the record with the highest score in the format Tuple(matched_record, score, index)

    # If Name + Address score is greater than the Full Name score, return Name + Address Record
    if name_address__score > full_name__score:
        return (name_address__name, name_address__score, name_address__id)

    # Else return the Full Name record
    else:
        address = voter_records_2023_df.loc[full_name__id, 'Full Address']
        full_name_address = f"{full_name__name} {address}"
        return (full_name_address, full_name__score, full_name__id)

Reply:
This has the same structure as the original code. I previously found it confusing because it had a series of `if` conditionals which had overlapping conditions and thus could be better served by being combined. In particular the final set of `if-else` conditionals should be a sub case of the `else` condition in the second `if-else` conditional since you only get to that final set if you don't `return` the second computation of `(name_address__name, name_address__score, name_addess__id)`. 

But I think I better understand what you're trying to do. I'm trying to maximize readability to make it easier to improve the code later on. Here is my previous code in a more readable form which also avoids the full voter records check if the second ward check satisfies the threshold. 

```
def tiered_search(name, address):

    name_address_combo = f"{name} {address}"

    # Searches for a match within the Ward returned by OCR
    name_address_matches1 = score_fuzzy_match_slim(name_address_combo, voter_records_2023_df[voter_records_2023_df['WARD'] == f"{dict_['Ward']}.0"]["Full Name and Full Address"])
    name_address__name1, name_address__score1, name_address__id1 = name_address_matches1[0]

    # if score is more than 85, return the tuple
    if name_address__score1 >= 85.0:
        return (name_address__name1, name_address__score1, name_address__id1)
    
    # if score is below 85, do additional processing
    else:

        # computing matches based on name and address; only considers all other wards
        name_address_matches2 = score_fuzzy_match_slim(name_address_combo, voter_records_2023_df[voter_records_2023_df['WARD'] != f"{dict_['Ward']}.0"]["Full Name and Full Address"])
        name_address__name2, name_address__score2, name_address__id2 = name_address_matches2[0]

        # if the new voter records score is greater than 85, return tuple
        if name_address__score2 >= 85.0:
            return (name_address__name2, name_address__score2, name_address__id2)
        
        # if score is less than 85, perform full records search based on name
        # and return results with highest score
        else: 
            # computing matches based on name alone; considers full voter records
            full_name_matches = score_fuzzy_match_slim(name, voter_records_2023_df["Full Name"], scorer_=fuzz.ratio)
            full_name__name, full_name__score, full_name__id = full_name_matches[0]

            # find max from three scores
            max_indx = np.argmax([name_address__score1, name_address__score2, full_name__score])

            # return records associated with that max
            if max_indx==0:
                return (name_address__name1, name_address__score1, name_address__id1)
            elif max_indx ==1: 
                return (name_address__name2, name_address__score2, name_address__id2)
            else: 
                address = voter_records_2023_df.loc[full_name__id, 'Full Address']
                full_name_address = f"{full_name__name} {address}"
                return (full_name_address, full_name__score, full_name__id)
```